task/distributed_queue.py

```diff
@@ -65,6 +66,10 @@
 }
 
 
+class DeclareResult:
```

Comment:
This still isn't setting well with me.

I looked into this a bit more, and the actual type of the thing here is a `pika.frame.Method`, but as implemented in `pika`, this isn't marked `Generic` so we have no way to instantiate it like `pika.frame.Method[pika.spec.Queue.DeclareOk]`.

There is, however, https://pypi.org/project/types-pika/ which does have a `Generic` `pika.frame.Method`: https://github.com/baod-rate/types-pika/blob/ccde5ce04818c1e678ff46f701203cd5322f2683/pika-stubs/frame.pyi#L21

I'm not sure it's worth including that as a dependency, though, and I'm not sure mypy would correctly deal with the case that the base library was present and the type annotations were missing.

The problem with what you're doing here is that PEP484 is nominal typing and you're trying to do structural typing here.  I think I'd feel a lot better about this if you subclassed it from `typing.Protocol`: https://docs.python.org/3/library/typing.html#typing.Protocol

Reply:
Wrt. your chat question:

> did you fail to get the type for the message_count field without doing the protocol for yourself? it seems like the kind of thing that mypy might have succeeded at scraping out of the source, even without proper types

I tried this:
```diff
-- task/distributed_queue.py
+++ task/distributed_queue.py
@@ -23,7 +23,7 @@ import os
 import ssl
 from collections.abc import Collection
 from types import TracebackType
-from typing import Any, Protocol, Self
+from typing import Any, Protocol, Self, TypeVar
 
 no_amqp = False
 try:
@@ -66,14 +66,13 @@ arguments = {
 }
 
 
-# conceptually these two are https://pika.readthedocs.io/en/stable/_modules/pika/spec.html#Queue.DeclareOk
+# conceptually this is https://pika.readthedocs.io/en/stable/_modules/pika/spec.html#Queue.DeclareOk
 # but we don't want to depend on types-pika, and are also only interested in one field
-class _PikaMethod(Protocol):
-    message_count: int
+T = TypeVar("T")
 
 
-class DeclareResult(Protocol):
-    method: _PikaMethod
+class DeclareResult(Protocol[T]):
+    method: T
```

and it doesn't work:
```
task/distributed_queue.py:130: error: Missing type parameters for generic type "DeclareResult"  [type-arg]
run-queue:106: error: Missing type parameters for generic type "DeclareResult"  [type-arg]
run-queue:107: error: Missing type parameters for generic type "DeclareResult"  [type-arg]
run-queue:108: error: Missing type parameters for generic type "DeclareResult"  [type-arg]
```

i.e. maybe I can specify some instance of T at the declaration, but IMHO that would be besides the point. So you most likely meant something else, but wihtout specifying `_PikaMethod`, what would `method`'s type be?

Reply:
This is all awkward, and it's also pointed out by the impedance mismatch in the [test failure](https://github.com/cockpit-project/bots/actions/runs/8818339405/job/24206896506?pr=6266). Some things like `run-queue` just blindly assume that all queues exist, but `DistributedQueue` cannot ensure that. This needs more thinking and bigger redesign. I'll back out the top (run-queue) commit for now, and work on that separately.

Reply:
The correct type, if we could, would be 
`pika.frame.Method[pika.spec.Queue.DeclareOk]`.

Of course, we can't.  So, first attempt:

```python
from typing import Protocol, reveal_type

import pika.spec


class DeclareFrame(Protocol):
    method: pika.spec.Queue.DeclareOk


y: DeclareFrame = None  # type: ignore[assignment]
reveal_type(y.method.message_count)
```

pyright is smart enough to figure out that this is `int`, but mypy says `Any`.  So much for that.

So I'd go with this instead:

```python
from typing import Protocol, reveal_type


class DeclareOk(Protocol):
    message_count: int


class DeclareFrame(Protocol):
    method: DeclareOk


y: DeclareFrame = None  # type: ignore[assignment]
reveal_type(y.method.message_count)
```

and then they both agree that this is an `int`.